strawberryfields/backends/tfbackend/circuit.py

```diff
@@ -709,140 +671,132 @@ def measure_homodyne(self, phi, mode, select=None, **kwargs):
+            eqn_lhs = batch_index + meas_mode_indices + "," + batch_index + conditional_indices
+            eqn_rhs = ''
+            meas_ctr = 0
+            cond_ctr = 0
+            for m in range(self._num_modes):
+                if m == mode:
+                    # use measured_indices
+                    eqn_rhs += meas_mode_indices[mode_size * meas_ctr : mode_size * (meas_ctr + 1)]
+                    meas_ctr += 1
                 else:
-                    # calculate norm of conditional_state
-                    # cheap hack since tensorflow doesn't allow einsum equation for trace:
-                    r = conditional_state
-                    for _ in range(self._num_modes - 2):
-                        r = ops.partial_trace(r, 0, False, self._batched)
-                    norm = tf.trace(r)
-
-                # for broadcasting
-                norm_reshape = [1] * len(conditional_state.shape[batch_offset:])
-                if self._batched:
-                    norm_reshape = [self._batch_size] + norm_reshape
-
-                normalized_conditional_state = conditional_state / tf.reshape(norm, norm_reshape)
-
-                # reset measured modes into vacuum
-                meas_mode_vac = self._single_mode_pure_vac if self._state_is_pure else self._single_mode_mixed_vac
-                batch_index = indices[:batch_offset]
-                meas_mode_indices = indices[batch_offset : batch_offset + mode_size]
-                conditional_indices = indices[batch_offset + mode_size : batch_offset + mode_size * self._num_modes]
-                eqn_lhs = batch_index + meas_mode_indices + "," + batch_index + conditional_indices
-                eqn_rhs = ''
-                meas_ctr = 0
-                cond_ctr = 0
-                for m in range(self._num_modes):
-                    if m == mode:
-                        # use measured_indices
-                        eqn_rhs += meas_mode_indices[mode_size * meas_ctr : mode_size * (meas_ctr + 1)]
-                        meas_ctr += 1
-                    else:
-                        # use conditional indices
-                        eqn_rhs += conditional_indices[mode_size * cond_ctr : mode_size * (cond_ctr + 1)]
-                        cond_ctr += 1
-                eqn = eqn_lhs + "->" + batch_index + eqn_rhs
-                new_state = tf.einsum(eqn, meas_mode_vac, normalized_conditional_state)
+                    # use conditional indices
+                    eqn_rhs += conditional_indices[mode_size * cond_ctr : mode_size * (cond_ctr + 1)]
+                    cond_ctr += 1
+            eqn = eqn_lhs + "->" + batch_index + eqn_rhs
+            new_state = tf.einsum(eqn, meas_mode_vac, normalized_conditional_state)
 
-                self._update_state(new_state)
+            self._update_state(new_state)
 
-            return meas_result
+        return meas_result.numpy()
```

Comment:
In the previous version it was a float (if I remember correctly the tests require that).

Reply:
I think this type was previously kept flexible, i.e., whether it was a tensor or an array was determined by the `eval` argument. See `meas_result = homodyne_sample.eval(feed_dict, session)` on line 782 of previous version..

Most tests use `eval=True`, so they expect a numpy array. The tests in `test_tf_symbolic` would have checked the other cases, where a `tf.Tensor` was returned.

Obviously the old tf symbolic stuff has gone by the wayside. But what are the consequences (in particular I'm thinking for differentiability) of casting to numpy explicitly here? Maybe there are none, since the measurement result being stochastic precludes the worry about it needing to be differentiable?

Reply:
I'm a bit worried about adding `.numpy()` within the backend itself. All stuff related to types (numpy arrays or tf tensors) should be in the user's hands. Previously we allowed the measurement results to be returned as symbolic tensors (when the user asked for it). I'd like to keep that functionality for tf2 (except the tensors would no longer be symbolic by default). 

If a test is failing, it is likely because it was written with different assumptions in mind. I suspect it might even be a bug or implicit assumption in the `Parameters` class (which was not written with tf2 in mind). In that case, it's better to let the test fail as an indication that we need to fix/revisit it, rather than forcing the backend itself to have a more restricted behaviour.